### PR TITLE
Add new modules documentation

### PR DESCRIPTION
**Pull request by matir91**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,10 +41,17 @@ Basically, you could manage all data streams with a `Listing` component, which i
...
+1. [**Network support:**](#network-support) Provides a `Listing` based on a common Retrofit service implementation.
+Note entities won't be saved in memory nor disk.
+1. [**Cache + Network support:**](#cache--network-support) Provides a `Listing` with cache support using a common Retrofit service implementation, and a [`DataSource`](https://developer.android.com/reference/android/arch/paging/DataSource) for caching the data.
 We recommend you use [Room](https://developer.android.com/topic/libraries/architecture/room) to provide the `DataSource`, because it will be easier. However, you could use any other `DataSource`.
 
+Fountain supports 2 types of Retrofit service adapters:
+- A [RxJava2 retrofit adapter.](https://github.com/square/retrofit/tree/master/retrofit-adapters/rxjava2)
+- A [Coroutine retrofit adapter.](https://github.com/JakeWharton/retrofit2-kotlin-coroutines-adapter)
+
+It also supports not using an adapter, you can also work with a simple Retrofit [call](https://square.github.io/retrofit/2.x/retrofit/retrofit2/Call.html).
```
Comment (FerrariBruno) [inline README.md]: Maybe `It also supports not using any of them, as you could work with a simple...`??

File: README.md [hunk 2]
```diff
@@ -63,40 +77,93 @@ dependencies {
 ## Usage
 You can read the [full documentation](https://xmartlabs.gitbook.io/fountain/).
 
+### Factory constructors
+There's one static factory object class for each each dependency.
+- FountainCoroutines: Used to get a [Listing] from a Retrofit service which uses a Coroutine adapter.
+- FountainRetrofit: Used to get the a [Listing] from a Retrofit service without using an special adapter.
```
Comment (FerrariBruno) [inline README.md]: should be `Used to get a` as in the previous line

File: README.md [hunk 3]
```diff
@@ -63,40 +77,93 @@ dependencies {
...
+- FountainRx: Used to get a [Listing] from a Retrofit service which uses a RxJava2 adapter.
+
+Each static factory have two constructor, one for each Fountain mode.
```
Comment (FerrariBruno) [inline README.md]: `Each static factory has two constructors, one for each Fountain mode`

File: README.md [hunk 4]
```diff
@@ -63,40 +77,93 @@ dependencies {
...
+
 ### Network support
 
-The `Listing` with network support can be obtained using:
+The `Listing` with network support can be obtained invoking `createNetworkListing` from the static factory class.
+
+It require only one argument, a `NetworkDataSourceAdapter`, which provides all operations that the library will use to handle the paging.
```
Comment (FerrariBruno) [inline README.md]: `It requires...`

File: README.md [hunk 5]
```diff
@@ -63,40 +77,93 @@ dependencies {
...
+  val pageFetcher: PageFetcher
+
+  @CheckResult
+  fun canFetch(page: Int, pageSize: Int): Boolean
+}
 ```
 
-There's only one required structure, `NetworkDataSourceAdapter<out ListResponse<Value>>`, which this library uses to handle the paging.
-There are two methods: one to check if a page could be fetched and another one to fetch it.
+`PageFetcher` is an structure which provides a way to fetch a page from a service call.
```
Comment (FerrariBruno) [inline README.md]: `is a structure`

Comment (FerrariBruno) [reply README.md]: `that provides a way`

File: README.md [hunk 6]
```diff
@@ -63,40 +77,93 @@ dependencies {
...
+`ListResponseWithPageCount`: Used when the service provides the amount of pages.
+
+```kotlin
+interface ListResponseWithPageCount<T> : ListResponse<T> {
+  fun getPageCount(): Long
+}
+```
+
+If you use either `ListResponseWithPageCount` or `ListResponseWithEntityCount` you can convert a `PageFetcher` to a `NetworkDataSourceAdapter`.
+That means that if the response has the number of pages or entities you can get a `NetworkDataSourceAdapter` without implement the `canFetch` method.
```
Comment (FerrariBruno) [inline README.md]: `without implementing`

File: README.md [hunk 7]
```diff
@@ -123,6 +190,28 @@ This will be executed in a transaction.
...
 This will be executed in a transaction.
 
+### Caching strategy
+The pagination strategy that **Fountain** is using can be seen in the following image:
+<br> <p align="center"> <img src="images/paginationStrategy.png" /> </p>
+
+It starts with an initial service data request.
+By default the initial data requested is three pages, but this value can be changed, in the [`PagedList.Config`](https://developer.android.com/reference/android/arch/paging/PagedList.Config.html), using the [`setInitialLoadSizeHint`](https://developer.android.com/reference/android/arch/paging/PagedList.Config.html#initialLoadSizeHint) method.
+This parameter can be set in the factory constructor method. 
+When the service data comes, all data is refreshed in the `DataSource` using the `CachedDataSourceAdapter`.
```
Comment (FerrariBruno) [inline README.md]: `When the data comes from the service`??

File: fountain-coroutines/src/main/java/com/xmartlabs/fountain/coroutines/adapter/NetworkDataSourceAdapter.kt [hunk 8]
```diff
@@ -18,11 +20,5 @@ interface CoroutinePageFetcher<T> {
...
-/** It is used to handle the paging state */
-interface CoroutineNetworkDataSourceAdapter<T> {
-  val coroutinePageFetcher: CoroutinePageFetcher<T>
-
-  /** Returns `true` if the page [page] with a size [pageSize] can be fetched */
-  @CheckResult
-  fun canFetch(page: Int, pageSize: Int): Boolean
-}
+/**It's [NetworkDataSourceAdapter] based on a [CoroutinePageFetcher]. */
+interface CoroutineNetworkDataSourceAdapter<T : ListResponse<*>> : NetworkDataSourceAdapter<CoroutinePageFetcher<T>>
```
Comment (FerrariBruno) [inline fountain-coroutines/src/main/java/com/xmartlabs/fountain/coroutines/adapter/NetworkDataSourceAdapter.kt]: `It's a...`

File: fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/NetworkDataSourceAdapter.kt [hunk 9]
```diff
@@ -18,11 +20,5 @@ interface RetrofitPageFetcher<T> {
...
 
-/** It is used to handle the paging state */
-interface RetrofitNetworkDataSourceAdapter<T> {
-  val retrofitPageFetcher: RetrofitPageFetcher<T>
-
-  /** Returns `true` if the page [page] with a size [pageSize] can be fetched */
-  @CheckResult
-  fun canFetch(page: Int, pageSize: Int): Boolean
-}
+/**It's [NetworkDataSourceAdapter] based on a [RetrofitPageFetcher]. */
```
Comment (FerrariBruno) [inline fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/NetworkDataSourceAdapter.kt]: same

File: fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/RetrofitNetworkDataSourceAdapterFactory.kt [hunk 10]
```diff
@@ -63,26 +63,22 @@ object RetrofitNetworkDataSourceAdapterFactory {
...
 ) = RetrofitNetworkDataSourceAdapterFactory.fromTotalEntityCountListResponse(this, firstPage)
 
 /**
  * Provides a [RetrofitNetworkDataSourceAdapter] implementation of a [ListResponseWithEntityCount] response.
  * It is used when the service returns the page count in the response.
  *
- * @param Value The value that the service returns.
- * @param ListResponseValue The response type that the service returns.
- * @param pageFetcher It is used to fetch each page from the service.
+ * @param ServiceResponse The response type that the service returns.
```
Comment (FerrariBruno) [inline fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/RetrofitNetworkDataSourceAdapterFactory.kt]: maybe we can change all these for something like `@param ServiceResponse The response type returned by the service.`? What do you think?

File: fountain-rx2/src/main/java/com/xmartlabs/fountain/rx2/adapter/NetworkDataSourceAdapter.kt [hunk 11]
```diff
@@ -18,11 +19,5 @@ interface RxPageFetcher<T> {
...
 
-/** It is used to handle the paging state */
-interface RxNetworkDataSourceAdapter<T> {
-  val rxPageFetcher: RxPageFetcher<T>
-
-  /** Returns `true` if the page [page] with a size [pageSize] can be fetched */
-  @CheckResult
-  fun canFetch(page: Int, pageSize: Int): Boolean
-}
+/**It's [NetworkDataSourceAdapter] based on a [RxPageFetcher]. */
```
Comment (FerrariBruno) [inline fountain-rx2/src/main/java/com/xmartlabs/fountain/rx2/adapter/NetworkDataSourceAdapter.kt]: same

File: README.md [hunk 12]
```diff
@@ -54,7 +61,14 @@ repositories {
...
-    implementation 'com.github.xmartlabs:fountain:0.3.0'
+    // This dependency is required only if you want to use a Retrofit service without a special adapter. 
+    implementation 'com.github.xmartlabs.fountain:fountain-retrofit:0.3.0'
+
+    // This dependency is required only if you want to use a Coroutine retrofit adapter.
+    implementation 'com.github.xmartlabs.fountain:fountain-coroutines:0.3.0'
+
+    // This dependency is required only if you want to use a RxJava2 retrofit adapter.
+    implementation 'com.github.xmartlabs.fountain:fountain-rx2:0.3.0'
 }
```
Comment (bryant1410) [inline README.md]: What if we want to use multiple adapters? Isn't a core code that will be imported twice? Will there be a conflict then?

File: README.md [hunk 13]
```diff
@@ -123,6 +190,28 @@ This will be executed in a transaction.
...
 
+### Caching strategy
+The pagination strategy that **Fountain** is using can be seen in the following image:
+<br> <p align="center"> <img src="images/paginationStrategy.png" /> </p>
+
+It starts with an initial service data request.
+By default the initial data requested is three pages, but this value can be changed, in the [`PagedList.Config`](https://developer.android.com/reference/android/arch/paging/PagedList.Config.html), using the [`setInitialLoadSizeHint`](https://developer.android.com/reference/android/arch/paging/PagedList.Config.html#initialLoadSizeHint) method.
+This parameter can be set in the factory constructor method. 
+When the service data comes from the service, all data is refreshed in the `DataSource` using the `CachedDataSourceAdapter`.
+Note that the `Listing` component will notify that the data changed.
```
Comment (bryant1410) [inline README.md]: ~data~

File: README.md (hunk 12 above)
Comment (matir91) [reply README.md]: No, no problem with that. For example if you want to include push notifications and deeplinks in your app using firebase, both dependencies will include the firebase core dependency. Here is the same. makes sense?

Comment (bryant1410) [reply README.md]: > No, no problem with that. For example if you want to include push notifications and deeplinks in your app using firebase, both dependencies will include the firebase core dependency. Here is the same. makes sense?

Yeah. So either the three import a core library transitively (the core is not just code but a library so there is reuse and not duplication). Is the core the first one of the 3? Or the core is not listed?

Comment (matir91) [reply README.md]: Yep, exactly. The core is not listed, but you can add it using `implementation 'com.github.xmartlabs.fountain:fountain-core:0.X.Y'`. Furthermore, if you take a look at the other modules dependencies, the core is included as an [`api` dependency](https://github.com/xmartlabs/fountain/blob/b396ab3b6580fe4325c4d37463930c0364bb2b07/fountain-coroutines/build.gradle#L35), that means that you can use the core stuff without declaring that dependency.

Comment (matir91) [reply README.md]: I don't know, for me it doesn't make much sense. I mean, it's an internal module, for example if you want to include the firebase push notification module, the documentation say that you have to import [`implementation 'com.google.firebase:firebase-messaging:17.3.3'`](https://firebase.google.com/docs/cloud-messaging/android/client), it doesn't talk about the core module. However, I know that it exists!

File: README.md (hunk 13 above)
Comment (matir91) [reply README.md]: I'm sorry, I didn't understand it

Comment (bryant1410) [reply README.md]: Sorry, I meant: ~the~ data

File: README.md (hunk 12 above)
Comment (bryant1410) [reply README.md]: Alright!

**Pull request merged**